File: Programmers/[42893] 매칭 점수/KingDonggyu/42893.js
function setMatchingScore(pageinfo) {
  for (const url of pageinfo.keys()) {
    const { links, baseScore } = pageinfo.get(url);

    links.forEach((link) => {
      if (!pageinfo.has(link)) {
        return;
      }

      const targetPage = pageinfo.get(link);
      const linkScore = baseScore / links.length;

      targetPage.matchingScore += linkScore;
    });
  }
}

function getBaseScore(page, word) {
  word = word.toLowerCase();

  const searchWords = page.split(/[^a-z]/i).filter((pageWord) => {
    return pageWord.toLowerCase() === word;
  });

  return searchWords.length;
}

function getLinks(page) {
  const aTags = page.match(/<a href="\S*"/g);

  if (aTags === null) {
    return [];
  }

  const links = aTags.map(tag => {
    return tag.split('https://')[1].slice(0, -1);
  });

  return links;
}

function getUrl(page) {
  const metaTag = page.match(/<meta property="og:url" content="\S*/)[0];
  const url = metaTag.split('https://')[1].slice(0, -3);

  return url;
}

function solution(word, pages) {
  const pageinfo = new Map();

  pages.map((page, index) => {
    const url = getUrl(page);
    const links = getLinks(page);
    const baseScore = getBaseScore(page, word);

    pageinfo.set(url, {
      index, url, links, baseScore,
      matchingScore: baseScore
    });
  });

  setMatchingScore(pageinfo);

  let maxScore = 0;
  let answer = 0;

  for (const url of pageinfo.keys()) {
    const { index, matchingScore } = pageinfo.get(url);
    if (maxScore < matchingScore) {
      [maxScore, answer] = [matchingScore, index];
    }
  }

  return answer;
}

File: Programmers/[42893] 매칭 점수/Johoseong/42893.py
import re
from decimal import Decimal
def solution(word, pages):
    answer = 0
    word = word.lower()
    infos = dict()

    for i, page in enumerate(pages):
        page = page.replace('\n', ' ').lower()
        my = re.findall('<meta property="og:url" content="(\S+)"', page) # 현재 페이지 url
        links = re.findall('<a href="(https://[\S]*)"', page) # 외부 링크 url들
        basic_num = 0
        for f in re.findall(r'[a-zA-Z]+', page): # 기본 점수
            if f == word:
                basic_num += 1
        infos[my[0]] = [i, len(links), basic_num, 0, links] # [인덱스번호, 외부 링크 수, 기본점수, 매칭점수, 링크 url 정보]
    
    answer = -1
    index = -1
    for i in infos: # 링크 점수 구함
        for link in infos[i][4]:
            if link not in infos:
                continue
            infos[link][3] += (infos[i][2] / infos[i][1])
    for i in infos: # 매칭 점수 구해서 가장 큰 인덱스 찾음
        matching = infos[i][2] + infos[i][3]
        if matching > answer:
            answer = matching
            index = infos[i][0]
    return index